src/nupic/types/Sdr.hpp

```diff
@@ -230,12 +232,12 @@ class SparseDistributedRepresentation : public Serializable
     /**
      * @attribute dimensions A list of dimensions of the SDR.
      */
-    const vector<UInt> &dimensions = dimensions_;
+    const vector<UInt> &dimensions() const { return dimensions_; }
```

Comment:
I don't like this.  Why do you think this way is better?

Reply:
> The SDR already allows the copy constructor...

Without this change, I was not able to call 
`SDR tmp = SP::initMapPotential_()` which now returns SDR. This code is used in SP and that lead me to do that change.

Reply:
For this task I would recommend passing an `SDR&` to `initMapPotential` for returning data.  Even if you get the copy constructor to work for this use-case, it still is going to do a deep-copy of all of the SDR data.

Reply:
I have used above recommendation in initMapPotential. 

But I get other use-cases where the copy-constructor would be useful. 
The error is: 

```
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/test/unit/regions/BacktrackingTMRegionTest.cpp:490:67: error: use of deleted function ‘nupic::SparseDistributedRepresentation& nupic::SparseDistributedRepresentation::operator=(const nupic::SparseDistributedRepresentation&)’
dense = n1region1->getOutput("activeCells")->getData().getSDR();
^
In file included from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/ntypes/ArrayBase.hpp:46:0,
from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/ntypes/Array.hpp:168,
from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/ntypes/Value.hpp:38,
from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/regions/BacktrackingTMRegion.hpp:37,
from /home/mmm/devel/HTM/htm-community/nupic.cpp/src/test/unit/regions/BacktrackingTMRegionTest.cpp:47:
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/types/Sdr.hpp:124:7: note: ‘nupic::SparseDistributedRepresentation& nupic::SparseDistributedRepresentation::operator=(const nupic::SparseDistributedRepresentation&)’ is implicitly deleted because the default definition would be ill-formed:
class SparseDistributedRepresentation : public Serializable
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/types/Sdr.hpp:124:7: error: non-static reference member ‘const std::vector<unsigned int>& nupic::SparseDistributedRepresentation::dimensions’, can’t use default assignment operator
/home/mmm/devel/HTM/htm-community/nupic.cpp/src/nupic/types/Sdr.hpp:124:7: error: non-static reference member ‘const UInt& nupic::SparseDistributedRepresentation::size’, can’t use default assignment operator
```

code that triggers it: 

```
auto sdr = region.getSDR(); //this is ok
//do some checks on sdr
//other updates to region
sdr = region.getSDR(); //fails
```

The calling code can be found in 
6818ef0

Reply:
> The calling code can be found in

I could call 

```
auto dense = ..
auto dense1 = 
auto dense2 =
```
but I don;t find that an OK code

Reply:
Ok, I think I understand why this is happening.  I can add an operator overload for assignment.  It should call the method `SDR::setSDR` which copies an SDR's data.

Reply:
I made a pull request with this feature, see PR #326.